Parallel Build Failing on Distributed Filesystem

I'm trying to compile OpenBLAS 0.3.12 using GCC 10.2.0. 

When it gets to the linktest part it fails with unresolved symbols. And when I look in the .so file using 'nm' I can see a load of symbols are undefined.

perl ./gensymbol linktest  x86_64 _ 0 0 0 0 0 0 "" "" 1 0 1 1 1 1 > linktest.c
gcc -O2 -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=80 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.12\" -msse3 -mssse3 -msse4.1 -march=skylake-avx512 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -shared -o ../libopenblas_skylakexp-r0.3.12.so \
-Wl,--whole-archive ../libopenblas_skylakexp-r0.3.12.a -Wl,--no-whole-archive \
-Wl,-soname,libopenblas.so.0 -lm -lpthread -lgfortran -lm -lpthread -lgfortran
gcc -O2 -DMAX_STACK_ALLOC=2048 -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=80 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.12\" -msse3 -mssse3 -msse4.1 -march=skylake-avx512 -mavx2 -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -w -o linktest linktest.c ../libopenblas_skylakexp-r0.3.12.so -L/opt/software/base/gcc/10.2.0/lib/gcc/x86_64-pc-linux-gnu/10.2.0 -L/opt/software/base/gcc/10.2.0/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../../../lib64 -L/lib/../lib64 -L/usr/lib/../lib64 -L/opt/software/base/gcc/10.2.0/lib/gcc/x86_64-pc-linux-gnu/10.2.0/../../..  -lgfortran -lm -lquadmath -lm -lc   && echo OK.
/tmp/ccaHbXUr.o: In function `main':
linktest.c:(.text.startup+0xf88): undefined reference to `slagge_'
linktest.c:(.text.startup+0xf8f): undefined reference to `slagsy_'
linktest.c:(.text.startup+0xf96): undefined reference to `slahilb_'
linktest.c:(.text.startup+0xf9d): undefined reference to `slakf2_'
linktest.c:(.text.startup+0xfa4): undefined reference to `slaran_'
linktest.c:(.text.startup+0xfab): undefined reference to `slarge_'
linktest.c:(.text.startup+0xfb2): undefined reference to `slarnd_'

However, if I turn off parallel make then it builds fine.

I have 40 cores in my server so could there be an issue with the make dependencies causing the dynamic library to be build from the statis library before the static library has finished being built?

Here is the relevant part of the build log, which shows the list of symbols that are missing. Let me know, and I can include the entire log, if needed.

```
make[2]: Entering directory '/global/D1/homes/james/ex3modules/defq/1.0.0/src/openblas-0.3.12/exports'
perl ./gensymbol linktest  x86_64 _ 0 0 0 0 0 0 "" "" 1 0 1 1 1 1 > linktest.c
cc -O2 -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DDYNAMIC_ARCH -DNO_WARMUP -DMAX_CPU_NUMBER=256 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.12\" -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -shared -o ../libopenblas-r0.3.12.so \
-Wl,--whole-archive ../libopenblas-r0.3.12.a -Wl,--no-whole-archive \
-Wl,-soname,libopenblas.so.0 -lm -lgfortran -lm -lgfortran
cc -O2 -DMAX_STACK_ALLOC=2048 -DUSE_LOCKING -Wall -m64 -DF_INTERFACE_GFORT -fPIC -DDYNAMIC_ARCH -DNO_WARMUP -DMAX_CPU_NUMBER=256 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.12\" -UASMNAME -UASMFNAME -UNAME -UCNAME -UCHAR_NAME -UCHAR_CNAME -DASMNAME= -DASMFNAME=_ -DNAME=_ -DCNAME= -DCHAR_NAME=\"_\" -DCHAR_CNAME=\"\" -DNO_AFFINITY -I..  -w -o linktest linktest.c ../libopenblas-r0.3.12.so -L/cm/shared/apps/slurm/20.02.7/lib/x86_64-linux-gnu -L/cm/shared/apps/slurm/20.02.7/lib/../lib -L/cm/shared/apps/slurm/20.02.7/lib64/../lib -L/cm/shared/apps/slurm/20.02.7/lib64/../lib -L/usr/lib/gcc/x86_64-linux-gnu/7 -L/usr/lib/gcc/x86_64-linux-gnu/7/../../../../x86_64-linux-gnu/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/7/../../../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/7/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/cm/shared/apps/slurm/20.02.7/lib -L/cm/shared/apps/slurm/20.02.7/lib64 -L/cm/shared/apps/slurm/20.02.7/lib64/slurm -L/cm/shared/apps/slurm/20.02.7/lib64 -L/usr/lib/gcc/x86_64-linux-gnu/7/../../../../x86_64-linux-gnu/lib -L/usr/lib/gcc/x86_64-linux-gnu/7/../../..  -lgfortran -lm -lquadmath -lm -lc   && echo OK.
/tmp/cc3YEbzE.o: In function `main':
linktest.c:(.text.startup+0xf88): undefined reference to `slagge_'
linktest.c:(.text.startup+0xf8f): undefined reference to `slagsy_'
linktest.c:(.text.startup+0xf96): undefined reference to `slahilb_'
linktest.c:(.text.startup+0xf9d): undefined reference to `slakf2_'
linktest.c:(.text.startup+0xfa4): undefined reference to `slaran_'
linktest.c:(.text.startup+0xfab): undefined reference to `slarge_'
linktest.c:(.text.startup+0xfb2): undefined reference to `slarnd_'
linktest.c:(.text.startup+0xfb9): undefined reference to `slaror_'
linktest.c:(.text.startup+0xfc0): undefined reference to `slarot_'
linktest.c:(.text.startup+0xfc7): undefined reference to `slatm1_'
linktest.c:(.text.startup+0xfce): undefined reference to `slatm2_'
linktest.c:(.text.startup+0xfd5): undefined reference to `slatm3_'
linktest.c:(.text.startup+0xfdc): undefined reference to `slatm5_'
linktest.c:(.text.startup+0xfe3): undefined reference to `slatm6_'
linktest.c:(.text.startup+0xfea): undefined reference to `slatm7_'
linktest.c:(.text.startup+0xff1): undefined reference to `slatme_'
linktest.c:(.text.startup+0xff8): undefined reference to `slatmr_'
linktest.c:(.text.startup+0xfff): undefined reference to `slatms_'
linktest.c:(.text.startup+0x1006): undefined reference to `slatmt_'
linktest.c:(.text.startup+0x1b82): undefined reference to `dlagge_'
linktest.c:(.text.startup+0x1b89): undefined reference to `dlagsy_'
linktest.c:(.text.startup+0x1b90): undefined reference to `dlahilb_'
linktest.c:(.text.startup+0x1b97): undefined reference to `dlakf2_'
linktest.c:(.text.startup+0x1b9e): undefined reference to `dlaran_'
linktest.c:(.text.startup+0x1ba5): undefined reference to `dlarge_'
linktest.c:(.text.startup+0x1bac): undefined reference to `dlarnd_'
linktest.c:(.text.startup+0x1bb3): undefined reference to `dlaror_'
linktest.c:(.text.startup+0x1bba): undefined reference to `dlarot_'
linktest.c:(.text.startup+0x1bc1): undefined reference to `dlatm1_'
linktest.c:(.text.startup+0x1bc8): undefined reference to `dlatm2_'
linktest.c:(.text.startup+0x1bcf): undefined reference to `dlatm3_'
linktest.c:(.text.startup+0x1bd6): undefined reference to `dlatm5_'
linktest.c:(.text.startup+0x1bdd): undefined reference to `dlatm6_'
linktest.c:(.text.startup+0x1be4): undefined reference to `dlatm7_'
linktest.c:(.text.startup+0x1beb): undefined reference to `dlatme_'
linktest.c:(.text.startup+0x1bf2): undefined reference to `dlatmr_'
linktest.c:(.text.startup+0x1bf9): undefined reference to `dlatms_'
linktest.c:(.text.startup+0x1c00): undefined reference to `dlatmt_'
linktest.c:(.text.startup+0x289b): undefined reference to `clagge_'
linktest.c:(.text.startup+0x28a2): undefined reference to `claghe_'
linktest.c:(.text.startup+0x28a9): undefined reference to `clagsy_'
linktest.c:(.text.startup+0x28b0): undefined reference to `clahilb_'
linktest.c:(.text.startup+0x28b7): undefined reference to `clakf2_'
linktest.c:(.text.startup+0x28be): undefined reference to `clarge_'
linktest.c:(.text.startup+0x28c5): undefined reference to `clarnd_'
linktest.c:(.text.startup+0x28cc): undefined reference to `claror_'
linktest.c:(.text.startup+0x28d3): undefined reference to `clarot_'
linktest.c:(.text.startup+0x28da): undefined reference to `clatm1_'
linktest.c:(.text.startup+0x28e1): undefined reference to `clatm2_'
linktest.c:(.text.startup+0x28e8): undefined reference to `clatm3_'
linktest.c:(.text.startup+0x28ef): undefined reference to `clatm5_'
linktest.c:(.text.startup+0x28f6): undefined reference to `clatm6_'
linktest.c:(.text.startup+0x28fd): undefined reference to `clatme_'
linktest.c:(.text.startup+0x2904): undefined reference to `clatmr_'
linktest.c:(.text.startup+0x290b): undefined reference to `clatms_'
linktest.c:(.text.startup+0x2912): undefined reference to `clatmt_'
linktest.c:(.text.startup+0x33d1): undefined reference to `zlagge_'
linktest.c:(.text.startup+0x33d8): undefined reference to `zlaghe_'
linktest.c:(.text.startup+0x33df): undefined reference to `zlagsy_'
linktest.c:(.text.startup+0x33e6): undefined reference to `zlahilb_'
linktest.c:(.text.startup+0x33ed): undefined reference to `zlakf2_'
linktest.c:(.text.startup+0x33f4): undefined reference to `zlarge_'
linktest.c:(.text.startup+0x33fb): undefined reference to `zlarnd_'
linktest.c:(.text.startup+0x3402): undefined reference to `zlaror_'
linktest.c:(.text.startup+0x3409): undefined reference to `zlarot_'
linktest.c:(.text.startup+0x3410): undefined reference to `zlatm1_'
linktest.c:(.text.startup+0x3417): undefined reference to `zlatm2_'
linktest.c:(.text.startup+0x341e): undefined reference to `zlatm3_'
linktest.c:(.text.startup+0x3425): undefined reference to `zlatm5_'
linktest.c:(.text.startup+0x342c): undefined reference to `zlatm6_'
linktest.c:(.text.startup+0x3433): undefined reference to `zlatme_'
linktest.c:(.text.startup+0x343a): undefined reference to `zlatmr_'
linktest.c:(.text.startup+0x3441): undefined reference to `zlatms_'
linktest.c:(.text.startup+0x3448): undefined reference to `zlatmt_'
collect2: error: ld returned 1 exit status
Makefile:181: recipe for target '../libopenblas-r0.3.12.so' failed
make[2]: *** [../libopenblas-r0.3.12.so] Error 1
make[2]: Leaving directory '/global/D1/homes/james/ex3modules/defq/1.0.0/src/openblas-0.3.12/exports'
Makefile:116: recipe for target 'shared' failed
make[1]: *** [shared] Error 2
make[1]: Leaving directory '/global/D1/homes/james/ex3modules/defq/1.0.0/src/openblas-0.3.12'
makefiles/openblas-0.3.12.mk:53: recipe for target '/global/D1/homes/james/ex3modules/defq/1.0.0/pkgs/openblas-0.3.12/.pkgbuild' failed
make: *** [/global/D1/homes/james/ex3modules/defq/1.0.0/pkgs/openblas-0.3.12/.pkgbuild] Error 2
````

--- a/lapack-netlib/TESTING/MATGEN/Makefile
+++ b/lapack-netlib/TESTING/MATGEN/Makefile
@@ -66,6 +66,7 @@ ZMATGEN = zlatms.o zlatme.o zlatmr.o zlatmt.o \
 endif
 
 .PHONY: all
+.NOTPARALLEL:
 all: $(TMGLIB)
 
 ALLOBJ = $(SMATGEN) $(CMATGEN) $(SCATGEN) $(DMATGEN) $(ZMATGEN) \

